### PR TITLE
feat[venom]: add effects to instructions

### PR DESCRIPTION
**Pull request by charles-cooper**
### What I did
add effects to instructions, which is helpful for several different ongoing efforts

### Commit message

```
this commit adds an `effects.py` file to venom, which describes
the effects of opcodes. this is useful for several ongoing efforts,
including CSE elimination and DFT pass improvements.
```

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: vyper/venom/basicblock.py [hunk 1]
```diff
@@ -242,6 +243,12 @@ def is_volatile(self) -> bool:
     def is_bb_terminator(self) -> bool:
         return self.opcode in BB_TERMINATORS
 
+    def get_read_effects(self):
+        return effects.reads.get(self.opcode, ())
+
+    def get_write_effects(self):
+        return effects.reads.get(self.opcode, ())
```
Comment (harkal) [inline vyper/venom/basicblock.py:250]:
```suggestion
        return effects.writes.get(self.opcode, ())
```

File: vyper/venom/effects.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
+_ALL = ("storage", "transient", "memory", "immutables", "balance", "returndata")
+
+_writes = {
+    "sstore": "storage",
+    "tstore": "transient",
+    "mstore": "memory",
+    "istore": "immutables",
+    "call": _ALL,
+    "delegatecall": _ALL,
+    "staticcall": "memory",
```
Comment (HodanPlodky) [inline vyper/venom/effects.py:10]:
```suggestion
    "staticcall": ("memory", "returndata"),
```

Comment (charles-cooper) [reply vyper/venom/effects.py]: nice catch, thanks!

File: vyper/venom/effects.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    "call": ALL,
+    "delegatecall": ALL,
+    "staticcall": ALL,
+    "returndatasize": RETURNDATA,
+    "returndatacopy": RETURNDATA,
+    "balance": BALANCE,
+    "selfbalance": BALANCE,
+    "log": MEMORY,
+    "revert": MEMORY,
+    "return": MEMORY,
```
Comment (cyberthirst) [inline vyper/venom/effects.py:57]: return has an offset to memory, right? so in theory you might be forced to zero-extend the memory

Comment (cyberthirst) [reply vyper/venom/effects.py:57]: applies to other opcodes as well

File: vyper/venom/effects.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
+from enum import Flag, auto
+
+
+class Effects(Flag):
```
Comment (cyberthirst) [inline vyper/venom/effects.py:4]: modeling code (e.g., for code copy) isn't useful?

File: vyper/venom/effects.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    "invoke": ALL,  # could be smarter, look up the effects of the invoked function
+    "dloadbytes": MEMORY,
+    "returndatacopy": MEMORY,
+    "calldatacopy": MEMORY,
+    "codecopy": MEMORY,
+    "extcodecopy": MEMORY,
+    "mcopy": MEMORY,
+}
+
+reads = {
```
Comment (cyberthirst) [inline vyper/venom/effects.py:42]: why aren't create opcodes modeled here?

File: vyper/venom/effects.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+ALL = ~EMPTY
+STORAGE = Effects.STORAGE
+TRANSIENT = Effects.TRANSIENT
+MEMORY = Effects.MEMORY
+IMMUTABLES = Effects.IMMUTABLES
+BALANCE = Effects.BALANCE
+RETURNDATA = Effects.RETURNDATA
+
+
+writes = {
```
Comment (cyberthirst) [inline vyper/venom/effects.py:23]: selfdestruct?

File: vyper/venom/effects.py (hunk 4 above)
Comment (charles-cooper) [reply vyper/venom/effects.py:4]: It's not that useful, since it can't change during execution

Comment (charles-cooper) [reply vyper/venom/effects.py:4]: (Ay least that I know of)

File: vyper/venom/effects.py (hunk 3 above)
Comment (charles-cooper) [reply vyper/venom/effects.py:57]: do we really care about zero extension? i think gas is an effect we don't need to preserve

Comment (cyberthirst) [reply vyper/venom/effects.py:57]: what about `msize`?

Comment (charles-cooper) [reply vyper/venom/effects.py:57]: hmm that's true. but return can't be reordered past other instructions anyways

Comment (cyberthirst) [reply vyper/venom/effects.py:57]: you have the same for eg `mload`

**Review by cyberthirst — APPROVED** [review]

**Comment by HodanPlodky** [comment]
Have you considered the effect of logs on other logs ie. in CSE you cannot replace the log by another log or what would happened if you switch log order. Not sure if it should be modeled in this PR

**Comment by charles-cooper** [comment]
> Have you considered the effect of logs on other logs ie. in CSE you cannot replace the log by another log or what would happened if you switch log order. Not sure if it should be modeled in this PR

That's a good one! We definitely should not reorder logs.

**Pull request merged**